mitodl/mitxonline
# 1627: improve hubspot sync script to handle large numbers of users

#### Pre-Flight checklist

#### What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/1627

#### What's this PR do?

1. Creates a scheduled celery task to sync all MITx Online User records with HubSpot.  Performs upsert to HubSpot contacts.
2. Adds a new attribute, `hubspot_sync_datetime`, to the `User` model in MITx Online.  This is used to record the datetime at which the last create or update to the related HubSpot contact was made.  Default value is null.  This field is visible through Django Admin.
3. Adds logging to the `batch_update_hubspot_objects_chunked` method which is called by the task created in bullet #1 and the `sync_db_to_hubspot` management command.  The logging will display the HubSpot user ID's updated in each chunk as well as the percentage completed after each chunk.

#### How should this be manually tested?
**Test creating a new MITx Online user with a successful HubSpot call**
1. Configure HubSpot in your `.env`
2. Register a new user through MITx Online and verify, through Django Admin, that the user has a populated `hubspot_sync_datetime` value and an associated HubSpot object exists.

**Test creating a new MITx Online user with an unsuccessful HubSpot call**
1. Remove or comment out the HubSpot configuration in your `.env`
2. Register a new user through MITx Online and verify, through Django Admin, that the user does not have a populated `hubspot_sync_datetime` value and an associated HubSpot object does not exist.

**Test the sync_db_to_hubspot command**
For this test, you will refer to the user created during "Test creating a new MITx Online user with an unsuccessful HubSpot call"
1. Configure HubSpot in your `.env`
2. Run the `sync_db_to_hubspot` management command from the web container like so: `python manage.py sync_db_to_hubspot --users update`.
3. Verify that you see the new log messages described (they will display in the Celery container).
4. Verify, through Django Admin, that the user has a populated `hubspot_sync_datetime` value and an associated HubSpot object exists.

**Test the sync-hubspot-user-contacts scheduled celery task**
1. Remove or comment out the HubSpot configuration in your `.env`
2. Register a new user through MITx Online.
3. Stop your MITx Online instance.  
4. Add back or uncomment your HubSpot configuration in your `.env`.
5. Open `main/settings.py` and update the `sync-hubspot-user-contacts` entry to be:
```
    "sync-hubspot-user-contacts": {
        "task": "hubspot_sync.tasks.sync_all_contacts_with_hubspot",
        "schedule": crontab(
            minute="*",
            hour="*",
            day_of_week="*",
            day_of_month="*",
            month_of_year="*",
        ),
    },
```
6. Restart your MITx Online instance.
7. View the logs for the Celery container and verify that the `sync-hubspot-user-contacts` task runs and the new log messages are displayed.
8. Verify the user you created in step 2 now has a populated `hubspot_sync_datetime` value and an associated HubSpot object exists.

**Updating an existing user's profile information**
1. Use an existing MITx Online user who has a populated `hubspot_sync_datetime` value and an associated HubSpot object exists.
2. Login with that user and update their first name or any other profile information.
3. Verify that the `hubspot_sync_datetime` value has been updated to reflect a more recent datetime.
